### PR TITLE
Trials refactoring

### PR DESCRIPTION
**Pull request by Leowisd**
This fix https://github.com/oncokb/oncokb/issues/2215.

### CODE REVIEW

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+                }
+            }
+        }
+    }
+
+    private boolean isSpecialTumorType(String cancerType){
+        return SpecialTumorTypes.contains(cancerType);
+    }
+
+    private Trial[] getTrialsForSpecialCancerTye(JSONObject tumors, String cancerType){
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:278]: Why are we using array instead of arraylist?

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                }
+                if (isMatched){
+                    res.add(trial);
+                    break;
+                }
+            }
+        }
+    }
+
+    private boolean isSpecialTumorType(String cancerType){
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:274]: this is a dup code

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+            drugs[k] = new Drug();
+            drugs[k].setDrugName((String)drug.get("drugName"));
+            drugs[k].setNcitCode((String)drug.get("ncitCode"));
+        }
+        return drugs;
+    }
+
+    private Trial[] getTrialByTumorTreatment(Trial[] trials, String treatment){
+        List<Trial> res = new ArrayList<>();
+        String[] tempDrugName = treatment.split(",|\\+");
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:243]: why are we matching back slash?

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+                return new ResponseEntity<>(result, status);
+            }
+        }
+        return new ResponseEntity<>(result, status);
+    }
+
+    private Tumor getTumor(JSONObject jsonObject, String oncoTreeCode){
+        Tumor tumor = new Tumor();
+        if (jsonObject.containsKey(oncoTreeCode)){
+            JSONObject tumorObj = (JSONObject)jsonObject.get(oncoTreeCode);
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:183]: You can convert the json to model directly. https://stackoverflow.com/questions/24231223/how-can-i-cast-a-jsonobject-to-a-custom-java-class

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java (hunk 3 above)
Comment (Leowisd) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:243]: "\\"is java escape character, '+' needs escape character to be identify.

File: core/src/main/java/org/mskcc/cbio/oncokb/model/clinicalTrialsMathcing/CancerTypesQuery.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package org.mskcc.cbio.oncokb.model.clinicalTrialsMathcing;
+
+/**
+ * Created by Yifu Yao on 2020-9-11
+ */
+public class CancerTypesQuery {
+    private String[] cancerTypes;
```
Comment (zhx828) [inline core/src/main/java/org/mskcc/cbio/oncokb/model/clinicalTrialsMathcing/CancerTypesQuery.java:7]: List?

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        return tumor;
+    }
+
+    private List<Trial> getTrialByTumorTreatment(List<Trial> trials, String treatment){
+        List<Trial> res = new ArrayList<>();
+        String[] tempDrugName = treatment.split(",|\\+");
+        for (int i = 0; i < tempDrugName.length; i++){
+            tempDrugName[i] = tempDrugName[i].trim();
+        }
+        Set<String> drugsNames = new HashSet<>(Arrays.asList(tempDrugName));
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:185]: Arrays.stream(test.split(",|\\+")).map(item-> item.trim()).collect(Collectors.toSet())

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        getTrialByTumorDrugName(res, trials, drugsNames);
+        return res;
+    }
+
+    private void getTrialByTumorDrugName(List<Trial> res, List<Trial> trials, Set<String> drugsNames){
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:190]: This method seems not related to tumor?

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+                for (String val: drugsNames){
+                    if (!armDrugNames.contains(val)){
+                        isMatched = false;
+                        break;
+                    }
+                }
+                if (isMatched){
+                    res.add(trial);
+                    break;
+                }
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:207]: I assume this is arm.getDrugs().stream().map(Drug::getDrugName).collect(Collectors.toSet()).containsAll(drugsNames)

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+            }
+        }
+    }
+
+    private List<Trial> getTrialsForSpecialCancerTye(JSONObject tumors, String cancerType){
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:212]: getTrialsForSpecialCancerType

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        List<Trial> trials = new ArrayList<>();
+        Set<String> reMappedCancerTypes = getTumorNamesBySpecialTumorType(cancerType);
+        for (String type: reMappedCancerTypes){
+            trials.addAll(getTrialsByCancerType(tumors, type));
+        }
+        return trials;
+    }
+
+    private Set<String> getTumorNamesBySpecialTumorType(String cancerType){
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:221]: Why is it related to SpecialTumorType? Can you use getTumorTypeName in the TumorTypeUtils?

**Review by zhx828 — CHANGES_REQUESTED** [review]

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java (hunk 10 above)
Comment (Leowisd) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:221]: Here we try to find out names of all types whose tumorForm type equals to "Liquid" or "Solid" so we can get names of special tumor types like "All Liquid" and "All Solid".

**Review by zhx828 — CHANGES_REQUESTED** [review]

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+    final String s3AccessKey = PropertiesUtils.getProperties("aws.s3.accessKey");
+    final String s3SecretKey = PropertiesUtils.getProperties("aws.s3.secretKey");
+    final String s3Region = PropertiesUtils.getProperties("aws.s3.region");
+
+    @PremiumPublicApi
+    @ApiOperation("Return a list of trials using OncoTree Code and/or treatment")
+    @ApiResponses(value = { @ApiResponse(code = 200, message = "OK", response = Tumor.class),
+            @ApiResponse(code = 400, message = "Error", response = String.class) })
+    @RequestMapping(value = "/trials", produces = { "application/json" }, method = RequestMethod.GET)
+    public ResponseEntity<?> trialsMatchingGet(
```
[the remark this answers is not in the document]
Comment (Leowisd) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:72]: Because /trials without treatment parameter will return a tumor and /trials with treatment will return a list of trials. Any better idea to do it?

[the remark this answers is not in the document]
Comment (zhx828) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:72]: What's the use case returning tumor?

[the remark this answers is not in the document]
Comment (Leowisd) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:72]: @zhx828  Like this:
image

**Review by zhx828 — APPROVED** [review]
